### DISCUSSION TITLE
When reoptimising dispatched routes, how do I make sure vehicles return to the actual depot?

### DISCUSSION THREAD

**Discussion by bmorlo** [PyVRP/PyVRP#572] [Q&A] [answered]
Hey @leonlan,
I've decided to open a new discussion on a problem that I am currently facing, since I believe my question applies to many VRP variants.

**How do I make sure that in a solution (in which I specify the number of vehicles) all vehicles are utilized i.e. each vehicles visits at least one client?** Is there a parameter that allows me to specify for example a minimum utilized capacity per vehicle? Of course this would make the solutions less optimal, but for my application this would be important.

I'm still working with your proposed solution to the "flexible starting point" CVRP. Here, I need the vehicles to actually find their way home eventually. There should not be the possibility that the solver decides to "leave" a vehicle at its (flexible) starting location because he (or she) considers that current location a depot already.

Many thanks as always for your support!
Cheers

**Comment by N-Wouda** [comment, marked as answer]
Hi @bmorlo!

If I understand correctly your setting is one of reoptimisation of already dispatched routes, where the vehicles are on the way (i.e., all have a unique starting point) and need to eventually return to a single end depot. PyVRP doesn't really understand this, because it assumes all vehicles need to return to their starting location eventually - that's, after all, the "depot". A workaround is available for non-empty routes (which I think we discussed in https://github.com/PyVRP/PyVRP/discussions/553), but vehicles that are unused don't move, so that won't fly in general.

Until we support different start and end depots directly, I guess you'll need to make sure your vehicles are all used. A way to do so would be to introduce a required dummy client per vehicle at the vehicle's starting location. If you set the distance and duration from the vehicle's starting location to that dummy client to zero, and large values from all other locations, PyVRP's solver should then prefer to visit that location with the nearest vehicle, and then no vehicle routes are empty any more. That should solve the problem.

I don't know how well this actually works though. It is a bit of a workaround.

**Reply by bmorlo** [reply]
@N-Wouda Yes, that is indeed what I'm working on. Will try your proposed workaround today and benchmark it against OR Tools.

I was also thinking about actively contributing to PyVRP to make the start/end node separate since I believe that would streamline my code a lot. But I'm not so sure how to do it. Will have a look. I'm definitely not a professional programmer. What do you guys think?

Cheers

**Reply by N-Wouda** [reply]
Great! I think this should be a fairly easy addition. A start and end depot need to be added to the `VehicleType`, and then the `pyvrp.search.Route` needs to be updated a little bit as well to use those different depots. Everything else is already in place: `pyvrp.search` already uses a start and end depot, so the search procedures won't require updating.

The only difficulty I see is in `Solution`, which now needs to also track distance/duration etc. for what are essentially empty routes. We do not currently allow empty routes in the Solution at all, so I don't really know how to proceed there yet.

Let us know if the workaround works for you, or if you run into issues implementing it!

**Comment by bmorlo** [comment]
@N-Wouda Implemented your proposed workaround. So far, the solutions look good. I get the exact same total distance values as with OR routing (I always have both PyVRP and OR routing solving an instance in parallel). Let's see if I run into issues with larger instances.

Will have a look at the implementation for the start end depot seperation in the next weeks. But if the workaround works well, I will have to deprioritize this.

**Reply by bmorlo** [reply]
Edit: I am currently unsure what value is mathematically suitable to use for the "very large distance" to force the solver to not visit a wrong dummy client. Right now I'm using `np.max(distance_matrix) * (1 + vehicle_capacity)` and this works just fine, apparently. I did this because I believe that if the distance needed to reach the dummy in the wrong tour is higher than what is possible in a theoretical "super-tour", the solver will never incorporate a dummy client into a wrong tour.

I have also encountered a funny thing that could be my mistake of course: If the routes are not maxed out anymore since the number of locations to be visited is decreasing at the end of my vehicle (re-)routing campaign, PyVRP does still give me the same tours as OR routing but it yields slightly different objective values. Since you guys are the experts, do you have any idea why this could be the case?
Cheers!

**Reply by N-Wouda** [reply]
> I am currently unsure what value is mathematically suitable to use for the "very large distance" to force the solver to not visit a wrong dummy client. 

PyVRP internally uses a `MAX_VALUE` (available in `pyvrp.constants`) that's currently set to `1 << 52`. If you're using `pyvrp.Model`, no edge distance or duration may be set larger than this (the model complains if you try). It should thus provide a good upper bound. Alternatively, not adding edges you do not want to have in the model in the first place achieves the same thing: any 'missing' edges are set to this value in the distance and duration matrices passed to `ProblemData`.

> PyVRP does still give me the same tours as OR routing but it yields slightly different objective values. 

What's "OR routing" and what are "slightly different objective values"?

**Reply by bmorlo** [reply]
Hi Niels,
what exactly do you mean by "1 << 52"? I opted for not setting the edges that I do not want the solver to use. Thanks for the hint!

OR routing is the library/ the solver by google OR tools made specifically for VRPs etc.
I always run into higher total distances (my obj. val.) with PyVRP compared to OR routing when one of the vehicles has already returned to the main depot and the other vehicles are still in the field. I belive this is due to some error in my code. Will get back with an update on that, once I have fully understood what causes the issue.

**Reply by N-Wouda** [reply]
>what exactly do you mean by "1 << 52"? 

That's 1 bit shifted 52 positions to the left, which corresponds to $2^{52}$. 

>I opted for not setting the edges that I do not want the solver to use. Thanks for the hint!

No problem!

>I always run into higher total distances (my obj. val.) with PyVRP compared to OR routing when one of the vehicles has already returned to the main depot and the other vehicles are still in the field. I belive this is due to some error in my code. Will get back with an update on that, once I have fully understood what causes the issue.

Sounds good!

**Reply by bmorlo** [reply]
Update on my issue. I was able to fix it; however I do not know why it works.

I attached part of the code that I am running. It specifies the edges of the model. For some context: Before these lines, I basically store the dummy clients at the end of a coordinates list. I store the main depot right at the beginning of that list. Then come the vehicle starting points of this iteration. After that come the actual clients that I'm serving in this iteration and then, as mentioned, the dummy clients. Finally, I compute the distance matrix. I am still using the workaround from @leonlan. But now I additionally check for the _correct_ artificial-depot-to-dummy-client assignment for every vehicle (starting point). Here I do three things: 1) When a dummy client would be reached by anything else (client or artificial depot): Do not set this edge at all. 2) When a dummy client would be reached from the correct artificial depot (which is exactly what we are trying to force the solver to do), set the edge distance to zero. 3) When the end depot would be reached from a dummy client, also set this edge distance to zero.

No. 3 fixes my issues.

What I do not understand is why, by setting the edge distance from the dummy clients to the end depot to zero, I can force the solver to apparently prefer routes back to the end depot instead of e.g. "skipping" the end depot at first and visiting locations around it before then ending the tour at the end depot. With No. 3 in place, PyVRP delivers the same routes as OR routing (for smaller instances), no matter what. However, I am not sure if it should be like this. By setting the edges from a dummy-client to the depot to zero, I sort of fake a lower distance. This cannot be optimal, right? But OR routing delivers exactly these results. Maybe Google is using a similar code for their "separate start and end location" implementation...

What are your thoughts on this?

```ruby    for frm in pyvrp_model.locations:
        for to in pyvrp_model.locations:
            location_that_is_left_idx = int(frm.name.split("_")[0])
            location_that_is_visited_idx = int(to.name.split("_")[1])

            if location_that_is_visited_idx >= first_dummy_client_location_index:
                # A dummy client was reached.
                if location_that_is_left_idx < first_dummy_client_location_index:
                    # The dummy client was reached by the correct artificial depot.
                    if (location_that_is_left_idx + (
                            len(_full_coords_list_dummy_clients_pyvrp) - len(_starting_points_indices) - 1)
                            == location_that_is_visited_idx):
                        # Make it a no-brainer to immediately visit it. We need the additional - 1!
                        pyvrp_model.add_edge(frm, to, distance=0)
            else:
                # A normal client was reached.
                if location_that_is_left_idx >= first_dummy_client_location_index and location_that_is_visited_idx == 0:
                    # Set a distance from a dummy client to the depot to = 0. This resolves the issue of suboptimal
                    # tours when there are few locations left at the end. Now PyVRP always prefers to send a vehicle back 
                    # to the main depot and use a vehile that is already at the main depot to cover another location.
                    # TODO: This is not very clean! WHY DOES THIS EVEN WORK?
                    pyvrp_model.add_edge(frm, to, distance=0)
                else:
                    correct_distance = _full_distances_dummy_clients_pyvrp[location_that_is_left_idx][
                        location_that_is_visited_idx] if frm != to else 0
                    pyvrp_model.add_edge(frm, to, distance=correct_distance)```

**Reply by N-Wouda** [reply]
If a vehicle travels an empty route (i.e., goes directly from its starting depot to its end depot), does OR-Tools count that distance? Your example seems to suggest it does not, but I do not know much about how OR-Tools deals with this.

**Reply by bmorlo** [reply]
I agree with you, OR tools might actually not do this at all. But to me, it does not make much sense. So I guess I will remove this particular line in my code again, which mimics the OR behaviour. If I get an answer from the OR team, I'll post it here for completeness.

**Reply by bmorlo** [reply]
Hi again.

While the OR-tools guys have not yet responded sufficiently to my questions, I put some more thought into the issue. I believe that the problem lies somewhere else.

Now my question to you would be: Is there a way of changing a paramter in PyVRP such that the solver would prefer routes of equal capacity/length? The issue that I'm facing is that in some scenarios PyVRP prefers to have e.g. 4 of the 5 vehicles visit as many clients as they can whilst the remaining single vehicle is underutilized and only visits a single customer. Since I want my solutions to serve the customers as fast as possible, tours of equal length would be quite important, actually.

As always, thanks for your help.
Cheers!

**Reply by N-Wouda** [reply]
We don't have any way currently to actively optimise for balanced routes. PyVRP is somewhat biased to use as a few vehicles as it can, so I can imagine that that results in fairly unequal routes. You might be able to impose some sort of maximum distance or duration limit to ensure routes are somewhat evened out, but that is not the same as explicitly optimising for it.

> Since I want my solutions to serve the customers as fast as possible

This seems somewhat related to cumulative VRP (where you accumulate duration), but that is not something we currently support.
